File: IRCClient/src/main/java/net/darkmeow/irc/client/network/handle/HandleClientConnection.java
package net.darkmeow.irc.client.network.handle;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.channel.EventLoopGroup;
import net.darkmeow.irc.client.AttributeKeys;
import net.darkmeow.irc.client.network.IRCClientConnection;

import java.util.UUID;

public class HandleClientConnection extends ChannelInboundHandlerAdapter {

    public final IRCClientConnection client;

    public final EventLoopGroup group;

    public HandleClientConnection(IRCClientConnection client, EventLoopGroup group) {
        this.client = client;
        this.group = group;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        final UUID uniqueID = UUID.randomUUID();

        ctx.channel().attr(AttributeKeys.UUID).set(uniqueID);

        client.channel = ctx.channel();
        client.channelUniqueId = uniqueID;

        super.channelActive(ctx);
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        final UUID uniqueID = ctx.channel().attr(AttributeKeys.UUID).get();

        if (client.channelUniqueId == uniqueID && client.channel != null) {
            client.channel = null;
            client.channelUniqueId = null;

            new Thread(() -> {
                client.base.userManager.reset();

                group.shutdownGracefully().syncUninterruptibly();
                client.base.listenable.onDisconnect(
                    client.base.resultManager.disconnectType,
                    client.base.resultManager.disconnectReason,
                    client.base.resultManager.disconnectLogout
                );
            }).start();
        }

        ctx.channel().attr(AttributeKeys.UUID).set(null);

        super.channelInactive(ctx);
    }

}

File: IRCServer/src/main/kotlin/net/darkmeow/irc/network/handles/HandleClientConnection.kt
package net.darkmeow.irc.network.handles

import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.handler.codec.haproxy.HAProxyMessage
import net.darkmeow.irc.network.NetworkManager
import net.darkmeow.irc.utils.ChannelAttrUtils.getUniqueId
import net.darkmeow.irc.utils.ChannelAttrUtils.setAddress
import net.darkmeow.irc.utils.ChannelAttrUtils.setUniqueId
import java.net.InetSocketAddress
import java.util.*

class HandleClientConnection(private val manager: NetworkManager): ChannelInboundHandlerAdapter() {

    override fun channelActive(ctx: ChannelHandlerContext) {
        ctx.channel().also { channel ->
            UUID.randomUUID().also { uuid ->
                channel.setUniqueId(uuid)

                manager.clients[uuid] = channel

                channel.setAddress(
                    (channel.remoteAddress() as? InetSocketAddress)
                        ?.let { "${it.address.hostAddress}:${it.port}" }
                        ?: "unknown"
                )
            }
        }
        super.channelActive(ctx)
    }

    override fun channelInactive(ctx: ChannelHandlerContext) {
        ctx
            .channel()
            .getUniqueId()
            .takeIf { it != UUID(0L, 0L) }
            ?.also {
                manager.clients.remove(it)
            }

        super.channelInactive(ctx)
    }

    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        if (msg is HAProxyMessage) {
            ctx.channel().setAddress("${msg.sourceAddress()}:${msg.sourcePort()}")
            msg.release()
        } else {
            super.channelRead(ctx, msg)
        }
    }
}